File: com/as.tool/lua/script/can.py
__lic__ = '''
/**
 * AS - the open source Automotive Software on https://github.com/parai
 *
 * Copyright (C) 2015  AS <parai@foxmail.com>
 *
 * This source code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation; See <http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt>.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */
 '''
 
try:
    from .AS import *
except:
    from AS import *

__all__ = ['can_open','can_write','can_read','can_close','can_reset','DFTBUS','can_set_default']

__can__ = can()

default_busid = 0

DFTBUS = 0xdeadbeef

def can_set_default(busid):
    global default_busid
    print('change default bus from %s to %s'%(default_busid, busid))
    default_busid = busid

def can_open(busid,device,port,baudrate):
    if(busid == DFTBUS): busid = default_busid
    return __can__.open(busid,device.encode('utf-8'),port,baudrate)

def can_write(busid,canid,data):
    '''can request write on can bus <busid>'''
    if(busid == DFTBUS): busid = default_busid
    return __can__.write(busid, canid, len(data), bytearray(data))

def can_read(busid,canid):
    ''' can request read a can frame from <canid> queue of <busid>'''
    if(busid == DFTBUS): busid = default_busid
    result,canid,dlc,data= __can__.read(busid,canid)

    return result,canid,data

def can_close(busid):
    if(busid == DFTBUS): busid = default_busid
    return __can__.close(busid)

def can_reset(busid):
    if(busid == DFTBUS): busid = default_busid
    return __can__.reset(busid)

File: com/as.tool/cancasexl.access/Can.c
/**
 * AS - the open source Automotive Software on https://github.com/parai
 *
 * Copyright (C) 2015  AS <parai@foxmail.com>
 *
 * This source code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation; See <http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt>.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */
/* ============================ [ INCLUDES  ] ====================================================== */
#include "Can.h"
#include <windows.h>
#include "vxlapi.h"
/* ============================ [ MACROS    ] ====================================================== */
#define CAN_REPORT_ERROR_AND_EXIT(error)													\
	do	{																					\
		if(XL_SUCCESS != error)																\
		{																					\
			printf("CAN ERROR In %s line %d: %s\n",__FUNCTION__,__LINE__,					\
							xlGetErrorString(error));										\
			return E_NOT_OK;																\
		}																					\
	}while(0)
/* ============================ [ TYPES     ] ====================================================== */
/* ============================ [ DECLARES  ] ====================================================== */
/* ============================ [ DATAS     ] ====================================================== */
/* ============================ [ LOCALS    ] ====================================================== */
static boolean 		IsXLReady = FALSE;
static XLportHandle xlPortHandle[32];
static XLaccess     xlAccess[32];
static uint64       xlPortInUseMask=0;
/* ============================ [ FUNCTIONS ] ====================================================== */
Std_ReturnType Can_Init(uint32 port,uint32 baudrate)
{
	char userName[32];
	XLstatus status;
	XLaccess accessMask;
	if(FALSE == IsXLReady)
	{
		status = xlOpenDriver();

		CAN_REPORT_ERROR_AND_EXIT(status);

		IsXLReady = TRUE;
	}


	sprintf(userName,"port%d",(int)port);
	accessMask = 1<<port;
	status= xlOpenPort(&xlPortHandle[port],userName,accessMask,&xlAccess[port],512,XL_INTERFACE_VERSION,XL_BUS_TYPE_CAN);
	CAN_REPORT_ERROR_AND_EXIT(status);

	xlPortInUseMask |= accessMask;

	status = xlCanSetChannelBitrate(xlPortHandle[port],xlAccess[port],baudrate);
	CAN_REPORT_ERROR_AND_EXIT(status);


	status = xlActivateChannel(xlPortHandle[port],xlAccess[port],XL_BUS_TYPE_CAN,XL_ACTIVATE_RESET_CLOCK);
	CAN_REPORT_ERROR_AND_EXIT(status);

	return E_OK;
}

Std_ReturnType Can_Receive(uint32 port,char** message)
{
	XLstatus status;
	XLaccess accessMask;
	accessMask = 1<<port;

	if(xlPortInUseMask&accessMask)
	{
		unsigned int EventCount = 1;
		XLevent Event;
		status = xlReceive(xlPortHandle[port],&EventCount,&Event);
		CAN_REPORT_ERROR_AND_EXIT(status);
		*message = xlGetEventString(&Event);
		CAN_REPORT_ERROR_AND_EXIT(status);
	}
	else
	{
		*message = "CAN ERROR: this port is not in active status\n";
		return E_NOT_OK;
	}

	return E_OK;
}

Std_ReturnType Can_Write(uint32  port, Can_PduType *pduInfo )
{
	XLstatus status;
	XLaccess accessMask;
	accessMask = 1<<port;

	if(xlPortInUseMask&accessMask)
	{
		XLevent Event;
		unsigned int EventCount = 1;
		Event.tag=XL_TRANSMIT_MSG;
		Event.tagData.msg.id=pduInfo->id;
		Event.tagData.msg.flags=0;
		assert(pduInfo->length<=8);
		memcpy(Event.tagData.msg.data,pduInfo->sdu,pduInfo->length);
		Event.tagData.msg.dlc=pduInfo->length;
		status = xlCanTransmit(xlPortHandle[port],xlAccess[port], &EventCount, &Event);
		CAN_REPORT_ERROR_AND_EXIT(status);
	}
	else
	{
		return E_NOT_OK;
	}

	return E_OK;
}

Std_ReturnType Can_DeInit(uint32 port)
{
	XLstatus status;
	XLaccess accessMask;
	accessMask = 1<<port;

	if(xlPortInUseMask&accessMask)
	{
		status = xlClosePort(xlPortHandle[port]);
		CAN_REPORT_ERROR_AND_EXIT(status);
		xlPortInUseMask &= ~accessMask;
	}

	if(0==xlPortInUseMask)
	{
		status = xlCloseDriver();
		CAN_REPORT_ERROR_AND_EXIT(status);
	}

	return E_OK;
}
